nibabel/cmdline/diff.py

```diff
@@ -0,0 +1,185 @@
+            return True
+        elif isinstance(value0, np.ndarray):
+            if np.any(value0 != value):  # special test for ndarray
+                return True
+            else:
+                return False
+
+        elif value0 != value:
+            return True
+
+    return False
+
+
+def get_headers_diff(file_headers, names=None):
+    """Get difference between headers
+
+    Parameters
+    ----------
+    file_headers: list of actual headers (dicts) from files
+    names: list of header fields to test
+
+    Returns
+    -------
+    dict
+      str: list for each header field which differs, return list of
+      values per each file
+    """
+    difference = OrderedDict()
+    fields = names
+
+    if names is None:
+        fields = file_headers[0].keys()
+
+    # for each header field
+    for field in fields:
+        values = [header.get(field) for header in file_headers]  # get corresponding value
+
+        # if these values are different, store them in a dictionary
+        if are_values_different(*values):
+            difference[field] = values
+
+    return difference
+
+
+def get_data_md5sums(files):
+
+    md5sums = [
+        hashlib.md5(np.ascontiguousarray(nib.load(f).get_data(), dtype=np.float32)).hexdigest()
+        for f in files
+    ]
+
+    if len(set(md5sums)) == 1:
+        return []
+
+    return md5sums
+
+
+def main():
+    """Getting the show on the road"""
+
+    parser = get_opt_parser()
+    (opts, files) = parser.parse_args()
+
+    nibabel.cmdline.utils.verbose_level = opts.verbose
+
+    assert len(files) >= 2, "Please enter at least two files"
+
+    if nibabel.cmdline.utils.verbose_level < 3:
+        # suppress nibabel format-compliance warnings
+        nib.imageglobals.logger.level = 50
+
+    file_headers = [nib.load(f).header for f in files]
+
+    if opts.header_fields:  # will almost always have a header field
+        # signals "all fields"
+        if opts.header_fields == 'all':
+            # TODO: header fields might vary across file types, thus prior sensing would be needed
+            header_fields = file_headers[0].keys()
+        else:
+            header_fields = opts.header_fields.split(',')
+
+    diff = get_headers_diff(file_headers, header_fields)
+    data_diff = get_data_md5sums(files)
+
+    if data_diff:
+        diff['DATA(md5)'] = data_diff
+
+    if diff:
+        print("These files are different.")
+        print("{:<15}".format('Field'), end="")
+
+        for f in files:
+            output = ""
+            i = 0
+            while i < len(f):
+                if f[i] == "/" or f[i] == "\\":
+                    output = ""
+                else:
+                    output += f[i]
+                i += 1
+
+            print("{:<55}".format(output), end="")
+
+        print()
+
+        for key, value in diff.items():
+            print("{:<15}".format(key), end="")
```

Comment:
I would suggest capturing string lengths (15, 55) into variables, as hardcoding them is not a good idea, and having them in variables allows us to adapt for different environments. For example, for terminals with smaller widths etc..

Reply:
And eventually the table Formatter which was used by nib-ls could be used to make it all dynamic without hardcoring much... And then eventually pyout by @kyleam could be used for the most flexible tabulator ;-)
But we thought here to get it to some Bible state first and then improve upon